Integrate Apache Http client with WebClient

martin-tarjanyi opened the pull request:
This PR aims to integrate Apache HttpClient 5 (which supports reactive streams and became GA last month) with WebClient.

Questions
------
* Does Spring have some internal load tests which can be run the new implementation to see how it behaves under stress?
* Should we support some kind of resource sharing just like it's done for the netty and jetty implementations?

martin-tarjanyi committed:
Integrate Apache http client with WebClient

pivotal-issuemaster commented:
@martin-tarjanyi Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=24700)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-framework&pullRequestId=24700) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpResponse.java:
```diff
@@ -0,0 +1,111 @@
+					if (!this.rejectSubscribers.compareAndSet(false, true)) {
+						throw new IllegalStateException("The client response body can only be consumed once.");
+					}
+				})
```
I'm not really sure why it is needed, I've seen this in the netty implementation as well and this one didn't work without it, either.

rstoyanchev replied:
This can happen when using a status handler for example:
```java
client.get().url("...").retrieve().onStatus(...)
```
The status handler is supposed to read or drain the body, but we don't know if it has, so we subscribe anyway and drain if we succeed.

pivotal-issuemaster commented:
@martin-tarjanyi Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=24700)!

martin-tarjanyi committed:
Fix javadoc

martin-tarjanyi committed:
Minor code and doc polishment

martin-tarjanyi committed:
Merge remote-tracking branch 'remotes/origin/master' into webclient_apache

martin-tarjanyi committed:
Fix handling of multiple request cookies for apache http client

poutsma commented:
I will try and pick this up in the 5.3 timeline.

poutsma commented:
Thank you for submitting this PR. Could you please make sure that it conforms to our [Code Style](https://github.com/spring-projects/spring-framework/blob/master/CONTRIBUTING.md#source-code-style)? This makes it significantly easier for us to review the PR.

Specifically, please make sure that there is an [empty line](https://github.com/spring-projects/spring-framework/wiki/Code-Style#blank-lines) in between class declaration and fields, and two before constructors. Please refer to other classes in the Spring Framework to see how they are formatted.

martin-tarjanyi committed:
Conform to code style

martin-tarjanyi committed:
Merge remote-tracking branch 'remotes/origin/master' into webclient_apache

# Conflicts:
#	spring-webflux/src/test/java/org/springframework/web/reactive/function/client/WebClientIntegrationTests.java

martin-tarjanyi commented:
* Added the requested code styles changes.
* Reverted my change in WebClientIntegrationTests as @rstoyanchev made some refactoring there.
  * First test verified multiple request cookie handling and showed a potential bug in reactor-netty. This probably should be handled as a separate PR/issue: https://github.com/spring-projects/spring-framework/pull/24700/commits/df32d7f026941fff2dc376d62cceb3424cd34d1f#diff-e25ee214b9bfab881247f141b0ddb407R629-R632
  * Second test case verified response cookie handling: https://github.com/spring-projects/spring-framework/pull/24700/commits/14a63dbf76cce125194d05820246aae3f61501b1#diff-e25ee214b9bfab881247f141b0ddb407R650-R684 - I found this test valuable as cookie mapping was a bit trickier in the Apache implementation than in the others. Now, I'm a bit uncertain whether I should extend some exisitng test case with this or just create a new one.
  * Please, advise on these.

poutsma commented:
> * Added the requested code styles changes.

Great! Thanks.

>   * First test verified multiple request cookie handling and showed a potential bug in reactor-netty. This probably should be handled as a separate PR/issue: [df32d7f#diff-e25ee214b9bfab881247f141b0ddb407R629-R632](https://github.com/spring-projects/spring-framework/commit/df32d7f026941fff2dc376d62cceb3424cd34d1f#diff-e25ee214b9bfab881247f141b0ddb407R629-R632)

Indeed, I can duplicate this issue locally. Could you please file a separate issue against Spring Framework for this problem? Not sure if the reason is in our connector or in Reactor Netty itself, but we can migrate the issue if need be.

>   * Second test case verified response cookie handling: [14a63db#diff-e25ee214b9bfab881247f141b0ddb407R650-R684](https://github.com/spring-projects/spring-framework/commit/14a63dbf76cce125194d05820246aae3f61501b1#diff-e25ee214b9bfab881247f141b0ddb407R650-R684) - I found this test valuable as cookie mapping was a bit trickier in the Apache implementation than in the others. Now, I'm a bit uncertain whether I should extend some exisitng test case with this or just create a new one.

Let's keep this as is, i.e. a new test in `WebClientIntegrationTests`, for now, it seems like a useful test to have.

I am now going to take a closer look at the code, so review comments will follow.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/MonoFutureCallbackAdapter.java:
```diff
@@ -0,0 +1,52 @@
+ * @param <T> the result type
+ */
+class MonoFutureCallbackAdapter<T> implements FutureCallback<T> {
+
```
I don't think it's necessary to make this a top-level class, so let's make it an inner class of `ApacheClientHttpConnector` instead, as that's the only place it's used.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpRequest.java:
```diff
@@ -0,0 +1,142 @@
+ * @since 5.3
+ * @see <a href="https://hc.apache.org/index.html">Apache HttpComponents</a>
+ */
+class ApacheClientHttpRequest extends AbstractClientHttpRequest {
```
Rename to `HttpComponentsClientHttpRequest`.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpResponse.java:
```diff
@@ -0,0 +1,120 @@
+ * @since 5.3
+ * @see <a href="https://hc.apache.org/index.html">Apache HttpComponents</a>
+ */
+class ApacheClientHttpResponse implements ClientHttpResponse {
```
Rename to `HttpComponentsClientHttpResponse`.

poutsma reviewed (commented):
Thanks again for submitting this PR.

While going through the code, I have some comments. Can you please resolve them?

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpConnector.java:
```diff
@@ -0,0 +1,110 @@
+ * @since 5.3
+ * @see <a href="https://hc.apache.org/index.html">Apache HttpComponents</a>
+ */
+public class ApacheClientHttpConnector implements ClientHttpConnector {
```
I'd argue that `Apache` HTTP Client is more recognizable for users than `HttpComponents` HTTP client. Consistency is a strong point, though.

poutsma replied:
It's mostly for consistency, but I also think that Apache is too broad a term, covering everything from the original httpd to the Java code that used to be known as Jakarta (which includes the HTTP client). By prefixing it with Apache, we would rule out any other possible connectors coming from the Apache project.

ok2c replied:
@poutsma @martin-tarjanyi Is there any chance you could call those classes `ApacheHC*` instead of `HttpComponents*`. Or `ApacheHttpComponents*` worst case?

poutsma replied:
We have decided to stick with the `HttpComponents` prefix for the class names, mainly for consistency with existing code (i.e. `HttpComponentsClientHttpRequestFactory`), but also because we do not use the `Apache` prefix for other classes that offer support for Apache projects (i.e. `TomcatHeadersAdapter`, `CommonsLogWriter`, `GroovyMarkupView`, etc.)

martin-tarjanyi committed:
Rename Apache client components for consistency

martin-tarjanyi committed:
Polish HttpComponents WebClient implementation

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpRequest.java:
```diff
@@ -0,0 +1,142 @@
+				.forEach(entry -> entry.getValue().forEach(v -> this.httpRequest.addHeader(entry.getKey(), v)));
+
+		if (!this.httpRequest.containsHeader(HttpHeaders.ACCEPT)) {
+			this.httpRequest.addHeader(HttpHeaders.ACCEPT, ALL_VALUE);
```
There are a couple of WebClient integration test cases which expect this header to be present with `ALL_VALUE` when `Accept` header is not explicitly set by user.
See Jetty implementation for the same problem:
https://github.com/spring-projects/spring-framework/blob/dc6f63f610ab0278c8c9d2c5c7cbbebb9928ec91/spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java#L143-L145

martin-tarjanyi committed:
Re-add response cookie integration test for WebClient

martin-tarjanyi committed:
Use CookieStore to send cookies with HttpComponents client

martin-tarjanyi committed:
Move response body stream handling to getBody method

martin-tarjanyi committed:
Pass content encoding and type to ReactiveEntityProducer

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpRequest.java:
```diff
@@ -0,0 +1,142 @@
+				.map(HttpCookie::toString)
+				.collect(Collectors.joining("; "));
+
+		this.httpRequest.addHeader(HttpHeaders.COOKIE, cookiesString);
```
I switched to using CookieStore. 

I have some concerns, though. CookieStore contains both request and response headers and it can even contain cookies from previous requests (depending on how the user configured the HttpClientContext). So when I [parse response cookies](https://github.com/spring-projects/spring-framework/blob/3b032e287c842ad3aba1d5580758e81354b09fd3/spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpResponse.java#L81) I have no way to tell which cookies came from the response and which ones are from a different source, so the current implementation returns all of them which is probably not the right behaviour(?).

poutsma replied:
I understand your concern, but I think the potential for bugs in the former version was simply too high.

If the underlying implementation does not offer us a way to differentiate between various cookies, then so be it. If we document that behavior, then it's good enough for me.

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpRequest.java:
```diff
@@ -0,0 +1,142 @@
+
+	@Override
+	public Mono<Void> writeAndFlushWith(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+		return writeWith(Flux.from(body).flatMap(p -> p));
```
Likely not, but I'm not sure. Let me tag Apache contributors @ok2c and @rschmitt maybe they can help out here. 
So the question is: Is it possible to control when the reactive data producer flushes parts of the request body?
Also, folks, you are more than welcome to review this pull request if you have time.

ok2c replied:
@martin-tarjanyi I personally have little experience with reactive bindings but I hope @rschmitt can answer your question. What I could do though is to look at various ways of optimizing the implementation once it gets accepted.

martin-tarjanyi commented:
I've addressed most of the comments and answered where I have concerns.

poutsma commented:
Thank you. I have some further comments and changes I'd like to see, but I do not want to ask too much of you. The code is definitely in a good enough state now for me to take over.

Would you be willing to go through another round of comments? Or should I wrap it up?

martin-tarjanyi commented:
I'm happy to work on this, but if you think at this point it would be more efficient to take over then feel free to do so. Thanks for your review.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpConnector.java:
```diff
@@ -0,0 +1,169 @@
+			ReactiveResponseConsumer reactiveResponseConsumer =
+					new ReactiveResponseConsumer(new MonoFutureCallbackAdapter<>(sink));
+
+			this.client.execute(basicRequestProducer, reactiveResponseConsumer, context, null);
```
Passing null here for the 4th parameter callback seems odd, but I suppose it is OK, because the completion signal is also passed to the response consumer, and from there to the callback adapter. Correct?

martin-tarjanyi replied:
Yes, that's my understanding.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpRequest.java:
```diff
@@ -0,0 +1,152 @@
+				});
+	}
+
+	public HttpRequest getHttpRequest() {
+		return this.httpRequest;
+	}
+
+	public Flux<ByteBuffer> getByteBufferFlux() {
+		return this.byteBufferFlux;
+	}
+
+	public long getContentLength() {
+		return this.contentLength;
+	}
```
These getters should be removed, in favor of having a `toRequestProducer` method. See above.

poutsma reviewed (commented):
Hi and thanks again,

I have left some more comments. Please help me resolve them.

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpConnector.java:
```diff
@@ -0,0 +1,169 @@
+	private ReactiveEntityProducer createReactiveEntityProducer(HttpComponentsClientHttpRequest request,
+			@Nullable Flux<ByteBuffer> byteBufferFlux) {
+
+		if (byteBufferFlux == null) {
+			return null;
+		}
```
When you say empty body you mean an empty `Flux` as the byte buffer publisher?
I tried this approach but it made the client hang.

poutsma replied:
Ok, that's fine then. I guess Apache HC does accept an empty data producer, but not empty an empty publisher.

martin-tarjanyi committed:
Code review changes

martin-tarjanyi commented on spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpConnector.java:
```diff
@@ -0,0 +1,169 @@
+	public HttpComponentsClientHttpConnector(CloseableHttpAsyncClient client,
+			Supplier<? extends HttpClientContext> contextSupplier) {
+
+		this.dataBufferFactory = new DefaultDataBufferFactory();
```
I did this change but I wonder how you decide whether it should be customizable through constuctor vs setter.

poutsma replied:
We have a setter in the `JettyClientHttpConnector`, so a setter here is fine.

rschmitt commented:
This looks good, from the perspective of `httpcore5-reactive`. I'm pleased to see that this integrates so cleanly with your existing abstractions (e.g. [`ReactiveHttpInputMessage`](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/ReactiveHttpInputMessage.java)); it's pretty much exactly what we were going for.

poutsma commented:
This PR has now been merged. Thanks for contributing, @martin-tarjanyi! And thanks for your support, @ok2c and @rschmitt.
